Add experimental write support for Tables.jl-compatible tables

jkrumbiegel opened the pull request:
I've moved the PR from ReadStat over, slightly editing it so it fits this library's formatting better. But I haven't yet made any conceptual changes.

I think your suggestion from the issue to use ReadStatTable as the input type is good, because it might make it easiest to handle all metadata correctly. Most of the work would then move into conversion methods between tables.

The base method could call `convert(ReadStatTable, table)` or something to that effect, and if the user needs more control, they could handle conversion themselves.

I'm not sure yet, what the whole list of features is that should be supported, as the different file formats have different capabilities as far as I remember from months ago. Do you have an idea which metadata is most important? Do you ideally want to have lossless roundtripping for a read-write-read scenario?

junyuan-chen commented on test/runtests.jl:
```diff
@@ -24,7 +24,10 @@ const tests = [
-    include("$test.jl")
-    println("\033[1m\033[32mPASSED\033[0m: $(test)")
-end
+@testset "ReadStatTables" verbose=true begin
```
The `verbose` keyword is not in Julia v1.3? The v1.3 support could stay with the v0.2.x versions until v0.3.0 is reached for which v1.6 will be required.

junyuan-chen commented:
@jkrumbiegel I just added more wrappers. The missing ones are all about metadata. I made some other changes to make use of the enum types that are already defined.

The changes break the part in `writer.jl`. I will soon update that part. I guess it won't take too much time to finish the remaining part and I will let you write tests if that's fine.

jkrumbiegel commented:
All right, thank you. Let me know when you think it's ready for tests then, so that we don't write conflicting code on this branch.

junyuan-chen commented:
@jkrumbiegel The basic functionality for write support is ready now. The tests haven't been written yet. So, I imagine that you could write the tests while taking a look at the code to verify that things work as expected.

There are some features/cases that have not been handled yet. I think it is worthwhile to get them covered in a separate PR later. So, this PR focuses on the basic cases that are just implemented and I will get back to the remaining parts later. These cases are listed below:

1. Convert columns with data/time values back to integers/floats and set an appropriate variable format based on the file format to be written. Values of type `Date` or `DataTime` are not supported at this moment.
2. Handle columns of type `CategoricalArray` or `PooledArray` in a better way. Value labels could be generated automatically from these arrays and only the reference values will be written.
3. Support `READSTAT_STRING_REF` for very long strings.

jkrumbiegel commented:
I've returned from winter break and will have a look at this soon.

jkrumbiegel commented:
I have added basic roundtrip tests for all formats. There's a commented out section where I'm not sure why the column labels changed during the roundtrip. For table-level metadata, it seemed to me that only `file_label`, `table_name` and `notes` store user-supplied strings, so I added only those specifically when converting to ReadStatTable. We should document somewhere that these are special metadata keys which will be reflected in written output (depending on file format) and that other keys will be dropped.

junyuan-chen commented:
@jkrumbiegel Great! I will take a closer look at the code in a couple of days. Please feel free to add additional changes/tests. I will wrap things up by figuring out the missing test coverage and add a docstring for `writestat`.

jkrumbiegel commented:
Thank you, that sounds good.

jkrumbiegel commented:
Hi @junyuan-chen, I haven't had more changes I wanted to add, what's the current state on your side? I could also take a crack at writing the docstring if you want.

junyuan-chen commented:
@jkrumbiegel Something else happened (not about this package) and took me much more time to handle than what I initially thought and I have obviously delayed the work here. But, I do plan to get back here very soon (in a day). Sorry about the delay!

If you happen to have some time right now, please feel free to make some changes. Otherwise, I will get started anyway.

jkrumbiegel commented:
Thank you, and don't worry about it. I just wanted to bump this so we didn't forget about the PR, there are always lots of other things going on :)
I can add a first docstring version today.

junyuan-chen commented:
@jkrumbiegel The changes made here will show up in v0.2.3, but I will make one more PR before that.

jkrumbiegel commented:
Thank you!
